src/main/java/com/google/devtools/build/lib/rules/proto/ProtoToolchain.java

```diff
@@ -0,0 +1,58 @@
+// Copyright 2021 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.rules.proto;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.devtools.build.lib.analysis.FilesToRunProvider;
+import com.google.devtools.build.lib.analysis.RuleContext;
+import javax.annotation.Nullable;
+
+/** Toolchain for {@code proto_*} rules. */
+public class ProtoToolchain {
```

Comment:
Use @AutoValue with this, so that the class gets reasonable `equals()` and `hashCode()` behavior.

Is this eventually going to be a provider returned from a `proto_toolchain` rule? If so, I'd go ahead and rename it to `ProtoToolchainInfo` now (and also so that there is less confusion for the future `ProtoToolchainRule` and `ProtoToolchain` classes for that).

Reply:
The plan is to implement `proto_toolchain` directly in Starlark and have a factory method `fromStarlark(ToolchainInfo)` that checks that there are `compiler` and `compiler_options` fields of the expected type for usage from native code.

I was advised against creating `ProtoToolchainInfo` that extends from `ToolchainInfo` because Starlark doesn't have inheritance and C++/Java recently migrated away from doing so.